lib/gc/Transforms/IterativeTilingAndFusion.cpp

```diff
@@ -0,0 +1,853 @@
+          return options.comparerPipeLine.compare(rewriter, candidateA,
+                                                  candidateB, defOrUse);
+        });
+    // f. call tiling interface
+    FailureOr<scf::SCFFuseConsumerOfSliceResult> fusedResult =
+        scfX::tileAndFuseConsumerOfSlice(rewriter, bestCandidate);
+
+    if (succeeded(fusedResult)) {
+      fusedResultList.push_back(*fusedResult);
+      auto whileProducerOutOfLoopBlock =
+          [&fusedResult](LoopLikeOpInterface loop) -> LogicalResult {
+        Block &body = loop->getRegion(0).front();
+        return failure(fusedResult.value().tiledOps[0]->getBlock() == &body);
+      };
+      SmallVector<LoopLikeOpInterface> outerLoops =
+          scfX::getOuterNestLoopsWhile(
+              bestCandidate->getParentOfType<LoopLikeOpInterface>(),
+              whileProducerOutOfLoopBlock);
+      // g. Manually run cse on region which contains top-level loop of
+      // candidate slice in avoid of conflict with subsequent
+      // `tileAndFuseConsumerOfSlice` get nest loops between next candidate
+      // sliceOp and tiled producer.
+      (void)mlir::simplifyRegions(rewriter,
+                                  {*outerLoops.front()->getParentRegion()});
+    }
+  }
+  if (fusedResultList.empty()) {
+    return std::nullopt;
+  } else {
+    return fusedResultList;
+  }
+}
+
+/// Target at following general topology:
+///
+/// producer1   producer2
+///    \         /
+///        Op
+///    /         \
+/// consumer1  consumer2
+///
+/// where:
+///
+/// Support iterative producer and consumer fusion in BFS fashion.
+LogicalResult iterativelyFuseProducerAndConsumerOfTiledOp(
+    RewriterBase &rewriter, Operation *tiledOp,
+    const CandidateSliceOptions &options) {
+  int numTiledOps = 0;
+  std::deque<Operation *> tiledOpList = {tiledOp};
+  while (!tiledOpList.empty()) {
+    tiledOp = tiledOpList.front();
+    tiledOpList.pop_front();
+    numTiledOps++;
+    // fuse producer
+    for (OpOperand &operand : tiledOp->getOpOperands()) {
+      if (std::optional<scf::SCFFuseProducerOfSliceResult> fuseProducerResult =
+              tileAndFuseProducerOfOpOperand(rewriter, operand, options))
+        tiledOpList.push_back(fuseProducerResult.value().tiledOps[0]);
+    }
+    // fuse consumer(s)
+    for (OpResult result : tiledOp->getResults()) {
+      if (std::optional<SmallVector<scf::SCFFuseConsumerOfSliceResult>>
+              fuseConsumerResults =
+                  tileAndFuseConsumerOfOpResult(rewriter, result, options)) {
+        for (auto &fuseConsumerResult : *fuseConsumerResults)
+          tiledOpList.push_back(fuseConsumerResult.tiledOps[0]);
+      }
+    }
+  }
+  return success(numTiledOps > 1);
+}
+
+/// What is single tiled op in loop?
+/// E.g.
+/// %1 = scf.for(){
+///   %2 = scf.for(){
+///       %3 = extract_slice
+///       %4 = tiled_op(%3)
+///       %5 = insert %4
+///       yield %5
+///   }
+/// }
```

Comment:
> This forbids the loop to have any regular linalg op except for the tiled one.

In fact, we are finding the self-tiled op which generates outer loop and candidate slice. Then we can call `iterativelyFuseProducerAndConsumerOfTiledOp` to fuse both producer and consumer by iterative way. 

If op is tiled by default upstream interface `scf::tileUsingSCF`, it is exactly matched with what we expect as above. If not, saying our manually deep tiled `matmul`,  we have to specially deal with them, like `linalg::FillOp` and `linalg::CopyOp` as you saw.

> What about a generic that is a broadcast?

In general, we suggest all other op except `matmul` should not be tiled together with `deep-tiled-matmul` before fusion unless it will become more complex during fusion stage. For example, if we have tiling on K-axis, `linalg::FillOp` should only be executed at the first K iteration. i.e. `k_i == 0`.

```
if(k_i == 0) {
  linalg.fill
}
brgemm()
```

And that is the reason why we bypass it at present. As for `broadcast`, I think it could be automatically fused by fusion pass rather than manually tiled before fusion. If there exist any other kinds of ops like `fill` on one day, perhaps we need to append them into filter as well...